Models/TreeModel.cpp

```diff
@@ -220,35 +245,80 @@ bool TreeModel::dropMimeData(const QMimeData *mimeData, Qt::DropAction action, i
       parents.remove(node);
       endRemoveRows();
     } else {
-      // duplicate the node
-      auto *dup = node->New();
-      dup->CopyFrom(*node);
-      node = dup;
-      // give the duplicate node a new name
-      const QString name = resourceMap->CreateResourceName(node);
-      node->set_name(name.toStdString());
-      // add the new node to the resource map
-      resourceMap->AddResource(node, resourceMap);
-    }
-
-    beginInsertRows(parent, row, row);
-    parentNode->mutable_child()->AddAllocated(node);
-    for (int j = parentNode->child_size() - 1; j > row; --j) {
-      parentNode->mutable_child()->SwapElements(j, j - 1);
+      node = duplicateNode(*node);
     }
-    parents[node] = parentNode;
-    endInsertRows();
 
-    ++row;
+    insert(parent, row++, node);
   }
 
   return true;
 }
 
-void TreeModel::addNode(buffers::TreeNode *child, buffers::TreeNode *parent) {
-  auto rootIndex = QModelIndex();
-  emit beginInsertRows(rootIndex, parent->child_size(), parent->child_size());
-  parent->mutable_child()->AddAllocated(child);
-  parents[child] = parent;
-  emit endInsertRows();
+QModelIndex TreeModel::addNode(buffers::TreeNode *child, const QModelIndex &parent) {
+  auto insertIndex = parent;
+  buffers::TreeNode *parentNode = nullptr;
```

Comment:
Please move this declaration into the parent.isValid branch. It's confusing to keep it around when it's barely ever applicable (it seems as though it's the parent you'll be inserting into, but it isn't used outside that branch except where an inline is preferable).

If it makes you feel better, rename `insertIndex` to `insertParent` or similar.

Reply:
Alright: 1ec23f6c28c5526fe55663c50cf89c5e6fcd2925

That seems reasonable enough.